Workaround to make PythonOps traced with torch.jit.trace work correctly.

zdevito opened the pull request:
The long-term fix is to remove the handling-creating pathways and
remove all the modes from PythonOp making it into an op that simply
calls a PyObject. Right now ONNX expects PythonOp to hold a
nn.Function, not a generic callable, so completely removing the legacy
pathway will also require changes to how ONNX symbolics are found.

zdevito committed:
Workaround to make PythonOps traced with torch.jit.trace work correctly.

The long-term fix is to remove the handling-creating pathways and
remove all the modes from PythonOp making it into an op that simply
calls a PyObject. Right now ONNX expects PythonOp to hold a
nn.Function, not a generic callable, so completely removing the legacy
pathway will also require changes to how ONNX symbolics are found.

zdevito commented:
@jamesr66a @apaszke

apaszke commented on test/test_jit.py:
```diff
@@ -1446,6 +1446,14 @@ def test_fuser_multiple_blocks(this, that, theother, meme):
+        @torch.jit.trace(torch.zeros(1, 2, 3))
+        def dropout_test(x):
+            x = torch.neg(x)
+            return F.dropout(x)
```
This test depends on `Dropout` being implemented in Python. Implement your own autograd function instead, or take the test from my PR.

apaszke commented on torch/csrc/jit/tracer_state.h:
```diff
@@ -55,6 +55,9 @@ struct TracingState : public std::enable_shared_from_this<TracingState> {
   std::mutex mutex;
   variable_list inputs; // Used only for the duration of first stage
 
+  bool creates_handles; // should python ops ever get handles? Or should
```
Why would this ever be `true`? When do we need it?

zdevito replied:
`torch.onnx.export` still uses the old tracing pathway where and we need do a bunch of yak shaving to remove it.

apaszke replied:
What do we need `creates_handles` for now? Looks like it's no longer necessary for ONNX, because it will correctly handle traces that had it set to `false`.

zdevito replied:
We need it because I am pretty sure the ONNX symbolics expect those handles to be there, and I want to keep that fix, which will take a lot of iteration with the onnx tests, as a separate pr.

ezyang replied:
I thought ONNX never supported handles at all, because it never supported backward export, and that's what the handles were originally added for. There is a bunch of special cases in the ONNX pass to make sure we handle handles correctly, but they're mostly checking if the handle is actually unused (which it should be if no backwards occurred.)

In any case if you guys need this to unblock I wouldn't block on this.

ezyang replied:
(But it would be nice to have a comment saying this is temporary, and solely for supporting ONNX)

apaszke commented on torch/csrc/jit/tracer.h:
```diff
@@ -184,8 +184,8 @@ inline Value* getOutputTrace(const std::shared_ptr<TracingState>& state, const V
 inline std::pair<std::shared_ptr<TracingState>, variable_list> enter(
-    variable_list inputs, std::size_t num_stages) {
-  auto state = std::make_shared<TracingState>(num_stages);
+    variable_list inputs, std::size_t num_stages, bool creates_handles) {
```
Just look how many places needed to be changed because the signature changed here. Since this is mostly a legacy thing, how about we remove it from there, and do a hacky `tracing_state->create_handles = true;` in those few places that need it.

zdevito commented:
Code is now updated. I had to change things around a bit because onnx export expects to be able to find ".symbolic" attributes, so we have to recover the original `Function` class in those cases. I added a `autogradFunction` object to do that.

apaszke commented on test/test_jit.py:
```diff
@@ -1446,6 +1446,41 @@ def test_fuser_multiple_blocks(this, that, theother, meme):
+        @torch.jit.trace(eg)
+        def foo(x):
+            x = torch.neg(x)
+            return F.dropout(x)
```
Can you please use a custom Python function instead of Dropout? It's going to break those tests once someone moves it to C++ and it's not their fault.

zdevito replied:
There are two tests here. One tests if dropout export works, which was the original reason for this PR. The other is a more generic test that tracing a Python function from a new trace works. I updated the names but both are useful tests.

apaszke commented on test/test_jit.py:
```diff
@@ -1446,6 +1446,41 @@ def test_fuser_multiple_blocks(this, that, theother, meme):
+        torch.onnx.export(MyDrop(), (eg,), f, verbose=False)
+
+    def test_python_function(self):
+        class MyLegacyFn(Function):
```
That's not a legacy function. Can you please change the name?

apaszke reviewed (commented):
Can you remove `creates_handles` now? Looks like we don't have the ONNX dependency anymore

ezyang reviewed (approved):
But it would be nice to add references in PythonOp to the explanation of what it contains.

ezyang commented:
Merge on green.
